mlir/lib/Target/LLVMIR/DebugImporter.cpp

```diff
@@ -25,6 +25,269 @@ using namespace mlir;
+      if (!unboundSelfRefs.empty())
+        translationStack.back().second.unboundSelfRefs.insert(
+            unboundSelfRefs.begin(), unboundSelfRefs.end());
+      return result;
+    }
+
+    // If the node type is capable of being recursive, check if it's seen
+    // before.
+    auto recSelfCtor = getRecSelfConstructor(node);
+    if (recSelfCtor) {
+      // If a cyclic dependency is detected since the same node is being
+      // traversed twice, emit a recursive self type, and mark the duplicate
+      // node on the translationStack so it can emit a recursive decl type.
+      auto [iter, inserted] = translationStack.try_emplace(node);
+      if (!inserted) {
+        // The original node may have already been assigned a recursive ID from
+        // a different self-reference. Use that if possible.
+        DIRecursiveTypeAttrInterface recSelf = iter->second.recSelf;
+        if (!recSelf) {
+          DistinctAttr recId = DistinctAttr::create(UnitAttr::get(context));
+          recSelf = recSelfCtor(recId);
+          iter->second.recSelf = recSelf;
+        }
+        // Inject the self-ref into the previous layer.
+        translationStack.back().second.unboundSelfRefs.insert(recSelf);
+        return cast<DINodeAttr>(recSelf);
+      }
+    }
+    return nullptr;
+  }
+
+  /// Register the translated result of `node`. Returns the finalized result
+  /// (with recId if recursive) and whether the result is self-contained
+  /// (i.e. contains no unbound self-refs).
+  std::pair<DINodeAttr, bool> finalizeTranslation(llvm::DINode *node,
+                                                  DINodeAttr result) {
+    // If `node` is not a potentially recursive type, it will not be on the
+    // translation stack. Nothing to set in this case.
+    if (translationStack.empty())
+      return {result, true};
+    if (translationStack.back().first != node)
+      return {result, translationStack.back().second.unboundSelfRefs.empty()};
+
+    TranslationState &state = translationStack.back().second;
+
+    // If this node is actually recursive, set the recId onto `result`.
+    if (DIRecursiveTypeAttrInterface recSelf = state.recSelf) {
+      auto recType = cast<DIRecursiveTypeAttrInterface>(result);
+      result = cast<DINodeAttr>(recType.withRecId(recSelf.getRecId()));
+
+      // Remove this recSelf from the set of unbound selfRefs.
+      state.unboundSelfRefs.erase(recSelf);
+
+      // Insert the newly resolved recursive type into the cache entries that
+      // rely on it.
+      // Only need to look at the caches at this level.
+      uint64_t numRemaining = state.cacheSize;
+      for (auto &cacheEntry : llvm::make_range(cache.rbegin(), cache.rend())) {
+        if (numRemaining == 0)
+          break;
+
+        if (auto refIter = cacheEntry.second.pendingReplacements.find(recSelf);
+            refIter != cacheEntry.second.pendingReplacements.end()) {
+          refIter->second = result;
+        }
+
+        --numRemaining;
```

Comment:
```suggestion
        if (numRemaining == 0)
          break;
        --numRemaining;

        auto refIter = cacheEntry.second.pendingReplacements.find(recSelf);
        if (refIter != cacheEntry.second.pendingReplacements.end()) 
          refIter->second = result;
```
nit: pure personal preference feel free to ignore :)